internal/pkg/vm/internal/vmcontext/validation/applier.go

```diff
@@ -0,0 +1,83 @@
+package validation
+
+import (
+	"context"
+
+	vtypes "github.com/filecoin-project/chain-validation/chain/types"
+	vstate "github.com/filecoin-project/chain-validation/state"
+	abi_spec "github.com/filecoin-project/specs-actors/actors/abi"
+
+	"github.com/filecoin-project/go-filecoin/internal/pkg/types"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/vm/actor/builtin"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/vm/internal/storage"
+	vmctx "github.com/filecoin-project/go-filecoin/internal/pkg/vm/internal/vmcontext"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/vm/state"
+)
+
+// Applier applies messages to state trees and storage.
+type Applier struct {
+}
+
+var _ vstate.Applier = &Applier{}
+
+func NewApplier() *Applier {
+	return &Applier{}
+}
+
+func (a *Applier) ApplyMessage(eCtx *vtypes.ExecutionContext, wrapper vstate.Wrapper, message *vtypes.Message) (vtypes.MessageReceipt, error) {
+	ctx := context.TODO()
+	st := wrapper.(*StateWrapper)
+
+	randSrc := &vmRand{}
+
+	tree, err := state.NewTreeLoader().LoadStateTree(ctx, st.storage, st.stateRoot)
+	if err != nil {
+		panic(err)
+	}
+	vmStrg := storage.NewStorage(st.bs)
+	gfcVm := vmctx.NewVM(randSrc, builtin.DefaultActors, &vmStrg, tree)
```

Comment:
Passing in default actors here forfeits the control chain-validation had over the Singleton actors it initializes, I don't think this is what we want but I lack a better solution.